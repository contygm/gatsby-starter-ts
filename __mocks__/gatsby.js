const React = require('react');
const gatsby = jest.requireActual('gatsby');

// NOTE: This mocks the graphql() function, Link component, and StaticQuery component.
// Allow testing components that use Link or GraphQL
module.exports = {
    ...gatsby,
    graphql: jest.fn(),
    Link: jest.fn().mockImplementation(
        // these props are invalid for an `a` tag
        ({
            activeClassName,
            activeStyle,
            getProps,
            innerRef,
            partiallyActive,
            ref,
            replace,
            to,
            ...rest
        }) =>
            React.createElement('a', {
                ...rest,
                href: to
            })
    ),
    StaticQuery: jest.fn(),
    useStaticQuery: jest.fn()
};
