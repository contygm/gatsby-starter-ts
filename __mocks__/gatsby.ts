import React from 'react';
const gatsby = jest.requireActual('gatsby');

// NOTE: This mocks the graphql() function, Link component, and StaticQuery component.
// Allow testing components that use Link or GraphQL
// https://stackoverflow.com/questions/54941990/how-to-strongly-type-jest-mocks
// https://stackoverflow.com/questions/48759035/mock-dependency-in-jest-with-typescript
module.exports = {
    ...gatsby,
    graphql: jest.fn(),
    Link: jest.fn().mockImplementation(
        // these props are invalid for an `a` tag
        ({
            activeClassName,
            activeStyle,
            getProps,
            innerRef,
            partiallyActive,
            ref,
            replace,
            to,
            ...rest
        }) =>
            React.createElement('a', {
                ...rest,
                href: to
            })
    ),
    StaticQuery: jest.fn(),
    useStaticQuery: jest.fn().mockImplementation(() => ({
        site: {
            siteMetadata: {
                title: `Starter`,
                siteUrl: `https://www.domain.com`,
                author: {
                    name: `Writer B. Author`,
                    description:
                        'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla quam velit, vulputate eu pharetra nec, mattis ac neque. Duis vulputate commodo lectus, ac blandit elit tincidunt id. Sed rhoncus, tortor sed eleifend tristique, tortor mauris molestie elit'
                },
                businessName: 'The Company',
                phone: '999-999-9999',
                email: 'email@email.com',
                description: 'a super duper site!',
                image: '/nothing.png'
            }
        }
    }))
};
