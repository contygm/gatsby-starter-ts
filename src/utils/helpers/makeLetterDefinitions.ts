import { GlossaryIndexProps } from '../../components';

/**
 * An object containing a letter and definitions that begin with that letter.
 *
 * @typedef {LetterWithDefinitions}
 *
 * @see GlossaryElements
 * @see makeLetterDefinitions
 */
type LetterWithDefinitions = {
    /** The letter all definitions begin with. */
    letter: string;
    /** An array of all the definitions that begin with the letter. */
    definitions: GlossaryElements[];
};

/**
 * A function that takes all the glossary definitions (sorted alphabetically) and splits the definitions based on the word's first letter.
 * The function returns an array of objects, each containing a letter and all definitions starting with that letter.
 *
 * @export
 * @function makeLetterDefinitions
 * @param {GlossaryIndexProps.allLetters} allLetters - All letters used by the glossary. Unused letters should not be included.
 * @param {GlossaryIndexProps.allDefinitions} allDefinitions - All definitions in the glossary. Needs to be sorted alphabetically.
 * @return {LetterWithDefinitions[]} - An array of objects with definitions sorted by letter
 *
 * @see GlossaryElements
 * @see GlossaryIndexProps
 */
export const makeLetterDefinitions = (
    allLetters: GlossaryIndexProps['allLetters'],
    allDefinitions: GlossaryIndexProps['allDefinitions']
): LetterWithDefinitions[] => {
    let index = 0;
    const resArray: LetterWithDefinitions[] = [];
    allLetters?.group.forEach((letterObj) => {
        const newLetterDefObj: LetterWithDefinitions = {
            letter: letterObj.fieldValue,
            definitions: []
        };

        while (
            index < allDefinitions.length &&
            letterObj.fieldValue === allDefinitions[index].frontmatter.letter
        ) {
            newLetterDefObj.definitions.push(allDefinitions[index]);
            index++;
        }
        resArray.push(newLetterDefObj);
    });
    return resArray;
};
