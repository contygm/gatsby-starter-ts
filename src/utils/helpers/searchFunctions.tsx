/**
 * A function that takes all the posts of a specific type and filters them based on a search query
 *
 * @function filterWithSearchQuery
 * @param {GlossaryElements[] | IndexElements[]} unfilteredPosts - All posts
 * @param {string} searchQuery - what to base the search on
 * @return {GlossaryElements[] | IndexElements[]} - includes only posts that match the search query
 *
 * @see GlossaryElements
 * @see IndexElements
 */
export const filterWithSearchQuery = (
    unfilteredPosts: (GlossaryElements | IndexElements)[],
    searchQuery: string
) => {
    if (searchQuery !== '') {
        const filteredData = unfilteredPosts.filter((post: any) => {
            const { title } = post.frontmatter;

            if ('description' in post.frontmatter) {
                // wiki or blog post
                const { excerpt } = post;

                return (
                    // standardize data with .toLowerCase()
                    // return true if the searchQuery is in the description or title
                    excerpt.toLowerCase().includes(searchQuery.toLowerCase()) ||
                    title.toLowerCase().includes(searchQuery.toLowerCase())
                );
            }

            const { html } = post;

            return (
                // standardize data with .toLowerCase()
                // return true if the searchQuery is in the definition or title
                html.toLowerCase().includes(searchQuery.toLowerCase()) ||
                title.toLowerCase().includes(searchQuery.toLowerCase())
            );
        });

        return filteredData;
    }

    return unfilteredPosts;
};
