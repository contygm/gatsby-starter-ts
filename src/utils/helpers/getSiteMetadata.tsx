import { useStaticQuery, graphql } from 'gatsby';

/**
 * a simple function to get all the site meta data from the `gatsby-config.ts` file
 * @function getSiteMetadata
 * @returns {SiteMetadata}
 */
export const getSiteMetadata = () => {
    const { site } = useStaticQuery(
        graphql`
            query SiteMetaData {
                site {
                    siteMetadata {
                        title
                        siteUrl
                        author {
                            name
                            description
                        }
                        businessName
                        phone
                        email
                    }
                }
            }
        `
    );
    return site.siteMetadata;
};
