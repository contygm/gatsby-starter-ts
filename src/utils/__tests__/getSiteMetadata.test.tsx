import { getSiteMetadata } from '../helpers/getSiteMetadata';
import * as gatsby from 'gatsby';

describe('getSiteMetadata', () => {
    it('return siteMetadata', () => {
        // NOTE: because this is mocked globally (__mocks__/gatsby), we give it a
        // different value to prove the hook will update based on siteMetadata
        const mock = jest.spyOn(gatsby, 'useStaticQuery');
        mock.mockImplementationOnce(() => ({
            site: {
                siteMetadata: {
                    title: `Different`,
                    siteUrl: `https://www.different.com`,
                    author: `Different B. Author`,
                    businessName: 'The Different Company',
                    phone: '111-111-1111',
                    email: 'different-email@email.com'
                }
            }
        }));

        const data = getSiteMetadata();
        expect(data).toMatchSnapshot();

        mock.mockRestore();
    });
});
