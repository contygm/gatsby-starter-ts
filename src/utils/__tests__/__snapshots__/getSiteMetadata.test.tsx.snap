// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSiteMetadata return siteMetadata 1`] = `
{
  "author": "Different B. Author",
  "businessName": "The Different Company",
  "email": "different-email@email.com",
  "phone": "111-111-1111",
  "siteUrl": "https://www.different.com",
  "title": "Different",
}
`;
